test/functional/eval/wait_spec.lua

```diff
@@ -0,0 +1,78 @@
+local helpers = require('test.functional.helpers')(after_each)
+local call = helpers.call
+local clear = helpers.clear
+local command = helpers.command
+local eval = helpers.eval
+local eq = helpers.eq
+local expect_err = helpers.expect_err
+local feed = helpers.feed
+local feed_command = helpers.feed_command
+local next_msg = helpers.next_msg
+local nvim = helpers.nvim
+local source = helpers.source
+
+before_each(function()
+  clear()
+  local channel = nvim('get_api_info')[1]
+  nvim('set_var', 'channel', channel)
+end)
+
+describe('wait()', function()
+  it('waits and returns 0 when condition is satisfied', function()
+    source([[
+    let g:_awake = 0
+    call timer_start(100, { -> nvim_command('let g:_awake = 1') })
+    ]])
+    eq(0, eval('g:_awake'))
+    eq(0, eval('wait(1500, { -> g:_awake })'))
+    eq(1, eval('g:_awake'))
+
+    eq(0, eval('wait(0, 1)'))
+  end)
+
+  it('returns -1 on timeout', function()
+    eq(-1, eval('wait(0, 0)'))
+    eq(-1, eval('wait(50, 0)'))
+  end)
+
+  it('returns -2 when interrupted', function()
+    feed_command('call rpcnotify(g:channel, "ready") | '..
+                 'call rpcnotify(g:channel, "wait", wait(-1, 0))')
+    eq({'notification', 'ready', {}}, next_msg())
+    feed('<c-c>')
+    eq({'notification', 'wait', {-2}}, next_msg())
+  end)
+
+  it('returns -3 on error', function()
+    command('silent! let ret = wait(-1, "error")')
+    eq(-3, eval('ret'))
+    command('let ret = 0 | silent! let ret = wait(-1, { -> error })')
+    eq(-3, eval('ret'))
+  end)
+
+  it('evaluates the condition on given interval', function()
+    source([[
+    function Count()
+      let g:counter += 1
+      return g:counter
+    endfunction
+    ]])
+
+    nvim('set_var', 'counter', 0)
+    eq(-1, call('wait', 20, 'Count() >= 5'))
```

Comment:
> Failure on AppVeyor, timing issue?

Weird, seems like it evaluated the condition at least five times in these 20 ms. My assumption here was that it's unlikely that the tests would raise like five events without any interaction, maybe increasing that `5` to something large (e.g.: `1000`) here and [there](https://github.com/neovim/neovim/pull/10382/files#diff-40765f886694def3efe2b0bd0ea82cf1R65) will do the trick?

Reply:
I am seeing a single event there locally only.
Maybe we should start by logging (unexpected) events then, if that's possible easily?  (does not look like it)

btw: the later tests should also be more strict maybe - it currently waits 1000ms, while 25ms should be enough.

Reply:
`multiqueue_put` is a macro. It could be possible to automatically log the name of every handler function using `#h` interpolation in the macro, at some high logging level. Alternatively log events as they are processed, but that would require shelling out to `llvm-symbolizer` to get the function name.

Reply:
> Alternatively log events as they are processed, but that would require shelling out to `llvm-symbolizer` to get the function name.

Interesting - how would that work then?  Do we use this somewhere already?

Reply:
Seems we use `addr2line` already in log.c which is the GNU equivalent (should work just as well)

Reply:
I'm unable to reproduce the failure though :|

Only with `eq(-1, call('wait', 20, 'Count() >= 2'))`:
```
INFO  2019-09-03T17:41:50.708 31709 main:564: starting main loop
INFO  2019-09-03T17:41:50.710 31709 multiqueue_process_events:154: calling request_event [0x55ac682a3650]
INFO  2019-09-03T17:41:50.710 31709 multiqueue_process_events:154: calling request_event [0x55ac682a3650]
INFO  2019-09-03T17:41:50.710 31709 multiqueue_process_events:154: calling request_event [0x55ac682a3650]
INFO  2019-09-03T17:41:50.711 31709 multiqueue_process_events:154: calling request_event [0x55ac682a3650]
INFO  2019-09-03T17:41:50.711 31709 f_wait:10988: evaluating condition: eval_expr_typval(&expr, &argv, 0, &exprval) != OK || tv_get_number_chk(&exprval, &error) || called_emsg || error || got_int
INFO  2019-09-03T17:41:50.730 31709 f_wait:10988: evaluating condition: eval_expr_typval(&expr, &argv, 0, &exprval) != OK || tv_get_number_chk(&exprval, &error) || called_emsg || error || got_int
WARN  2019-09-03T17:41:50.732 31709 call_set_error:613: RPC: ch 1 was closed by the client
INFO  2019-09-03T17:41:50.732 31709 multiqueue_process_events:154: calling exit_event [0x55ac682a20f0]
INFO  2019-09-03T17:41:50.732 31709 process_teardown:144: evaluating condition: kl_empty(loop->children) && multiqueue_empty(loop->events)
INFO  2019-09-03T17:41:50.732 31709 multiqueue_process_events:154: calling free_channel_event [0x55ac68140840]
INFO  2019-09-03T17:41:50.732 31709 process_teardown:144: evaluating condition: kl_empty(loop->children) && multiqueue_empty(loop->events)
INFO  2019-09-03T17:41:50.732 31709 mch_exit:99: Nvim exit: 0
```

Reply:
> > Failure on AppVeyor, timing issue?
> 
> Weird, seems like it evaluated the condition at least five times in these 20 ms. My assumption here was that it's unlikely that the tests would raise like five events without any interaction, ~maybe increasing that `5` to something large (e.g.: `1000`) here and [there](https://github.com/neovim/neovim/pull/10382/files#diff-40765f886694def3efe2b0bd0ea82cf1R65) will do the trick?~ (Oops, no that misses the point of the test, and the timeout will not be enough in first place)

I don't see why it misses the point of the tests?

Reply:
I forgot why I said that :)
Excuse my recent inactivity, busy with some stuff